[p-xp] iP

p-xp opened the pull request:
# Remy
> "The hand that mocked them, and the heart that fed" -- From Ozymandias by Percy Bysshe Shelley ([source](https://www.poetryfoundation.org/poems/46565/ozymandias))

Ever wanted to get insulted by a task manager that doesn't get much done at all? Remy is here to fill that niche.

What Remy does:
- _Add_ and _remove_ tasks from your tasklist
- Mark tasks as completed / not completed ✅❌
- [ ] Provides a good user experience

Want to get mocked? Here's the fastest way for that:
1. Run the application
2. Type a command that does not exist

---
You can launch Remy from the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

p-xp committed:
feat: implement Level-0

p-xp committed:
feat: implement Level-1

p-xp committed:
feat: implement Level-2

p-xp committed:
feat: implement Level-3

p-xp committed:
feat: implement Level-4

p-xp committed:
feat: implement TextUITesting

p-xp committed:
feat: implement Level-5

p-xp committed:
feat: implement Level-6

p-xp committed:
feat: implement Level-7

p-xp committed:
Merge pull request #1 from p-xp/branch-Level-7

feat: implement Level-7

p-xp committed:
add citations to Level-7

p-xp committed:
Merge pull request #2 from p-xp/branch-Level-7

add citations to Level-7

p-xp committed:
feat: implement date parsing

p-xp committed:
Merge pull request #3 from p-xp/branch-Level-8

feat: implement date parsing

p-xp committed:
feat: implement OOP

p-xp committed:
delete unused code

p-xp committed:
Merge pull request #4 from p-xp/branch-A-MoreOOP

implement OOP

p-xp committed:
feat: split into packages

p-xp committed:
Merge pull request #5 from p-xp/branch-A-Packages

feat: split into packages

p-xp committed:
Merge pull request #6 from p-xp/add-gradle-support

Add gradle support

p-xp committed:
feat: Enable Gradle

p-xp committed:
implement Gradle

p-xp committed:
Merge pull request #7 from p-xp/branch-A-Gradle

feat: Enable Gradle

p-xp committed:
feat: implement JUnit tests

p-xp committed:
Add citations

p-xp committed:
Merge pull request #8 from p-xp/branch-A-JUnit

implement JUnit testcases

p-xp committed:
create JAR file

p-xp committed:
Merge pull request #9 from p-xp/branch-A-Jar

create JAR file

p-xp committed:
added JavaDoc comments

p-xp committed:
tweak code to meet coding standard

p-xp committed:
implement find function

p-xp committed:
Merge pull request #10 from p-xp/branch-A-CodingStandard

tweak code to meet coding standard

p-xp committed:
implement find function

p-xp committed:
Merge branch 'master' into branch-A-JavaDoc

p-xp committed:
added JavaDoc comments

p-xp committed:
Fix bug

euchangxian commented on src/main/java/Remy/ChatbotException.java:
```diff
@@ -0,0 +1,21 @@
+    @Override
+    public String toString() {
+        return this.getMessage();
+    }
```
I noticed that the `toString()` method in `ChatbotException` has been overridden to return the error message. The default behaviour of the `Exception::toString` method already includes this message. Unless you meant to provide a custom behaviour, you should consider removing this overridden method to simplify the code!

euchangxian commented on src/main/java/Remy/Parser.java:
```diff
@@ -0,0 +1,52 @@
+import remy.command.MarkCommand;
+import remy.command.TodoCommand;
+import remy.command.UnmarkCommand;
+
+/**
+ * A class that makes sense of user inputs.
+ */
+public class Parser {
+
+    /**
+     * Parses user input and carries out any follow-up action (e.g. remy.remy.Task.remy.remy.Task creation).
```
I believe the extra `remy.`s in `remy.remy.Task` could be an unintentional mistake! Do consider reverting this change.

euchangxian commented on src/main/java/Remy/Parser.java:
```diff
@@ -0,0 +1,52 @@
+            return new FindCommand(input);
+        case ExitCommand.COMMAND_WORD:
+            return new ExitCommand();
+        default:
+            throw new ChatbotException("No such command lah.");
```
Nice use of the `default` branch! 👍

euchangxian commented on src/main/java/Remy/Remy.java:
```diff
@@ -0,0 +1,60 @@
+     * @param filePath location to store tasklist after ending each session.
+     */
+    public Remy(Path filePath) {
+        ui = new Ui();
```
I noticed `ui` is instantiated, but no instance methods of `ui` is invoked. Perhaps you might want to change the existing static methods to instance method. This will allow for easier unit testing of `Ui` as well, since static methods cannot be overridden to create stubs.

euchangxian commented on src/main/java/Remy/Ui.java:
```diff
@@ -0,0 +1,123 @@
+public class Ui {
+
+    public final Scanner scanner;
+    private static String divider = "____________________________________________________________\n";
+    private static String shortDivider = "_____________\n";
+    private static String welcomeMessage =
```
You could prefix `final` to these variables that I assume are meant to be constants? If so, do remember to change their names to `DIVIDER`, `SHORT_DIVIDER` and `WELCOME_MESSAGE` respectively to align with the coding standards.

euchangxian commented on src/main/java/Remy/Task/TaskList.java:
```diff
@@ -0,0 +1,82 @@
+    public String findMatchingTasks(String searchString) {
+        String matchingTasksString = "";
+        for (Task task : this.tasks) {
+            if (task.getDescription().toLowerCase().contains(searchString)) {
```
Good work converting the strings to lower case for case-insensitive comparison!👍

euchangxian commented on src/main/java/Remy/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,70 @@
+            Deadline temp = new Deadline(this.taskName, this.dueDate);
+            taskList.add(temp);
+            storage.save(taskList);
+            Ui.printAddedTask(temp, taskList.size());
```
Consider changing `Ui::printAddedTask` to an instance method instead!

euchangxian commented on src/main/java/Remy/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+     * Whether the Task should trigger the end of the current Chatbot session.
+     * @return boolean indicating Exit status.
+     */
+    public abstract boolean isExit();
```
This could possibly be a concrete method that returns `false` by default, and only overridden by `ExitCommand` to return `true`. This is so that not every subclass that extends `Command` have to override it, improving simplicity of your code.

euchangxian commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,29 @@
-|____/ \__,_|_|\_\___|
+____________________________________________________________
+
+I'm Remy.Remy, and it is NOT nice to see you.
```
Possibly an oversight here `Remy.Remy`. Consider reverting this change!

euchangxian commented on text-ui-test/runtest.bat:
```diff
@@ -15,7 +15,7 @@ IF ERRORLEVEL 1 (
 
 REM run the program, feed commands from input.txt file and redirect the output to the ACTUAL.TXT
-java -classpath ..\bin Duke < input.txt > ACTUAL.TXT
+java -classpath ..\bin remy.remy < input.txt > ACTUAL.TXT
```
Same issue as above!

euchangxian reviewed (commented):
Overall LGTM. Great implementation of the Commands class!

remuslum commented on src/main/java/Remy/command/FindCommand.java:
```diff
@@ -0,0 +1,42 @@
+        if (matcher.matches()) {
+            this.searchString = matcher.group(1).toLowerCase();
+        } else {
+            throw new ChatbotException("format wrong bro");
```
Perhaps on top of telling "bro" that his format is wrong, you could additionally provide the correct format to him so "bro" knows what he is doing wrong. E.g find ___ -> Did you forget to input a word?

remuslum commented on src/test/java/Remy/command/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,40 @@
+public class DeadlineCommandTest {
+    @Test
+    public void DeadLineCommand_missingInfo_throwChatbotException() {
+        assertThrows(ChatbotException.class, () -> new DeadlineCommand("deadline"));
```
I like the way you named your function to include the type of error like missing info and invalid date. To make it even better, might I suggest that instead of naming 1,2 and making it general, perhaps a more specific title could be used e.g invalid_date_missing_month, invalid_date_missing_year

remuslum commented on src/test/java/Remy/command/MarkCommandTest.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Test
+    public void MarkCommand_negativeIndex_throwChatbotException() {
```
Great job here. The titles here are not too general and can be easily comprehended 👍🏻

remuslum commented on src/main/java/DeleteCommand.java:
```diff
@@ -0,0 +1,26 @@
+        return false;
+    }
+
+    @Override
```
I don't think there is a need to provide an overwrite tag for the execute method as the method has not been implemented in the Command class.

p-xp committed:
Add checkstyle and removed styling violations

p-xp committed:
Add checkstyle and removed styling violations

p-xp committed:
Implement GUI using JavaFX

p-xp committed:
Implement GUI using JavaFX

p-xp committed:
Add assertions to Commands

There are no assertions in the code.

Adding assertions is ideal as it helps detect bugs and errors.

Let's add assertions that check the value of indexes stored in
Commands. This ensures that we do not attempt to execute an
invalid index.

p-xp committed:
Replace if-else with guard clauses

Commands use if-else blocks to check that index is valid.

If-else blocks can make code less readable as they obscure the
happy path. In contrast, guard clauses contribute to high code
quality as they make the happy path more prominent.

Let's replace all if-else checks with guard clauses.

p-xp committed:
Merge pull request #16 from p-xp/branch-A-CodeQuality

Replace if-else with guard clauses

p-xp committed:
Merge with master

p-xp committed:
Merge pull request #15 from p-xp/branch-A-Assertions

A-Assertions: Add assertions to Commands

p-xp committed:
Implement priority function for Deadline and Todo tasks

Users cannot assign priority to tasks.

This is an important functionality as users want to know what tasks they should focus on.

Let's implement a priority function so that users can indicate priority (high, medium, low)
for their tasks by typing /p.

p-xp committed:
Merge pull request #17 from p-xp/branch-BCD-Extension

Implement priority function for Deadline and Todo tasks

p-xp committed:
fixed bugs

p-xp committed:
Merge pull request #19 from p-xp/branch-BCD-Extension

fixed bugs

p-xp committed:
implement design changes

p-xp committed:
minor design changes

p-xp committed:
Wrote user guide

p-xp committed:
Fixed broken links

p-xp committed:
Created user guide

p-xp committed:
create JAR file

p-xp committed:
Merge pull request #22 from p-xp/branch-A-Release

create JAR file

p-xp committed:
Fixed delete command bug

p-xp committed:
Merge pull request #23 from p-xp/branch-A-Release

Fixed delete command bug
